src/projects/EnsureThat/EnsureOptions.cs

```diff
@@ -4,15 +4,22 @@ namespace EnsureThat
 {
     public struct EnsureOptions
     {
+        /// <summary>
+        /// If <see cref="CustomExceptionFactory"/> is defined, this exception will be thrown instead of the
+        /// standard exceptions for the particular ensure method.
+        /// Assign using <see cref="WithException(Exception)"/>.
+        /// </summary>
+        public Exception CustomException => CustomExceptionFactory(string.Empty);
+
         /// <summary>
         /// If defined, this exception will be thrown instead of the
         /// standard exceptions for the particular ensure method.
-        /// Assign using <see cref="WithException"/>.
+        /// Assign using <see cref="WithException(Func&lt;string,Exception&gt;)"/>.
         /// </summary>
-        public Exception CustomException { get; private set; }
+        public Func<string, Exception> CustomExceptionFactory { get; private set; }
```

Comment:
Right now this takes only the original message, right?
I feel like taking the paramname and value would be more important than preserving whatever text the library supplied out of box. Thoughts?

Reply:
Yes, it does indeed take only the original message.
Not sure if paramname and value gives any value. During the call to Ensure, I already know both of these. In that case, I could already do this, without this pull request, right?
`Ensure.String.HasLengthBetween(name, 0, 10, nameof(Name), o => o.WithException(new DomainValidationException($"The given string \"{name}\" is too long. Must be between '0' and '10'.", nameof(Name))));`
However, what I wanted to achieve is not to do anything, just to change the type of the exception, preserving the message.